lib/led-matrix/src/LEDMatrix.cpp

```diff
@@ -0,0 +1,154 @@
+/* LED Matrix Source File */
+
+// naming with m_
+#include "LEDMatrix.h"
+
+#include "mbed.h"
+
+// init thread ptr to nullptr and initialize the requestEndFlashing semaphore to 0 and a max of 1
+LEDMatrix::LEDMatrix(PinName R, PinName G, PinName B)
+    : m_RChannel(R), m_GChannel(G), m_BChannel(B), lightsThread(nullptr), reqEndFlash(0, 1) {}
+
+void LEDMatrix::terminateFlashing() {
+  reqEndFlash.release();  // ask flashing thread to end - signal semaphore
+  lightsThread->join();   // wait for the thread to finish. we've asked it to finish so should be quick wait
+  lightsThread->terminate();
+  delete lightsThread;
+  lightsThread = nullptr;
+}
+void LEDMatrix::flashing() {
+  while (!reqEndFlash.try_acquire()) {  // while we havent been asked to terminate this thread
+    printf("Flashing thread\r\n");
+    // do flashing with flashing_color which is a member var
+
+    while (true) {
+      setColor(flashing_red, flashing_green, flashing_blue);  // is okay that it calls the thread that tries to terminate it... nvmm
+      ThisThread::sleep_for(period_delay);
+      clearLights();
+      ThisThread::sleep_for(period_delay);
+    }
+  }
+}
+
+void LEDMatrix::setColor(uint8_t R, uint8_t G, uint8_t B) {
+  if (lightsThread == nullptr) {  // make sure a thread exists before trying to terminate it
+    terminateFlashing();
+  }
+
+  m_RChannel.pulsewidth(R / 255.0);
+  m_GChannel.pulsewidth(G / 255.0);
+  m_BChannel.pulsewidth(B / 255.0);
+}
+// Just for convenience. Call setColorRGB().
+void LEDMatrix::setColor(HWBRIDGE::LEDMATRIX::color c) {
+  if (lightsThread == nullptr) {
+    terminateFlashing();
+  }
+
+  switch (c) {
+    case HWBRIDGE::LEDMATRIX::color::RED:
+      setColor(255, 0, 0);
+      break;
+    case HWBRIDGE::LEDMATRIX::color::GREEN:
+      setColor(0, 255, 0);
+      break;
+    case HWBRIDGE::LEDMATRIX::color::BLUE:
+      setColor(0, 0, 255);
+      break;
+  }
+}
+
+void LEDMatrix::flashColor(uint8_t R, uint8_t G, uint8_t B) {
+  flashing_red = R;
+  flashing_green = G;
+  flashing_blue = B;
+
+  if(lightsThread == nullptr) { // make sure to not create a thread if another one is running
+    lightsThread = new Thread;
```

Comment:
heap allocations can fail. in which case they return nullptrs. u should check that new didnt return a nullptr bc currently ur code will crash. if new does nullprt, the function should just return;

Reply:
We should not be spawning new short lived threads like this. Especially in firmware. Since this library doesnt have a "run" function that gets called, a single thread should just be created at the constructor and reused